feat: Allow toggling autoclose

ditsuke opened the pull request:
## Summary
- Adds a command to allow toggling autoclose on the fly.
- Also gives autoclose the ability to close multiple buffers, to the point where we reach the open buffer threshold or exhaust closeable buffers. This is necessitated by the ability
  to toggle autoclose.

## Note
Please review #2 before this PR.

ditsuke committed:
feat(autoclose): Allow toggling with a command

- Adds a command to toggle auto close.
- Makes autoclose able to close multiple buffers in one
  trigger, which is majorly a necessity born out of the
  toggling feature that would let someone open >threshold
  buffers and then enable autoclose, which would mean
  the expected behavior is closing buffers until
  open buffers <= threshold or closable (non-pinned, live)
  buffers are exhausted.
- docs: Update README

ditsuke committed:
fix: exclude any buffer open in any window from autoclose

axkirillov commented on lua/hbac/command.lua:
```diff
@@ -1,26 +1,39 @@
+	if state.autoclose_enabled then
+		state.autoclose_enabled = false
+		pcall(function()
+			vim.api.nvim_del_augroup_by_name(require("hbac.setup").CONSTANTS.AUGROUP_AUTO_CLOSE)
```
Could you move this to a new module? Let's call it `autocommand.lua` and it will be responsible for creating / deleting autocommands.  That way `augroup_autoclose` becomes a local variable and you don't need to require setup here.

ditsuke replied:
Great idea! All done

axkirillov commented on lua/hbac/command.lua:
```diff
@@ -1,26 +1,39 @@
 end
 
+M.subcommands.toggle_autoclose = function()
+	if state.autoclose_enabled then
+		state.autoclose_enabled = false
+		pcall(function()
+			vim.api.nvim_del_augroup_by_name(require("hbac.setup").CONSTANTS.AUGROUP_AUTO_CLOSE)
+		end)
+		print("Hbac: Autoclose disabled")
+	else
+		state.autoclose_enabled = true
+		require("hbac.setup").setup_autoclose()
+		print("Hbac: Autoclose enabled")
+	end
```
```suggestion
	if state.autoclose_enabled then
	    -- if you want to run pcall you should pass the returned value here to handle it
	    if autocommand.remove_autoclose() then
		    state.autoclose_enabled = false
		    print("Hbac: Autoclose disabled")
		end
	else
		autocommand.add_autoclose()
		state.autoclose_enabled = true
		print("Hbac: Autoclose enabled")
	end
```

axkirillov commented on lua/hbac/setup.lua:
```diff
@@ -1,18 +1,75 @@
-		autoclose = true,
-		threshold = 10
-	}
+M.setup_autoclose = function()
```
this will then move to autocommand module and be called `add_autoclose`

axkirillov commented on lua/hbac/setup.lua:
```diff
@@ -22,65 +79,24 @@ M.setup = function(opts)
+		command.vim_cmd_func(args.args)
+	end, command.vim_cmd_opts)
 
 	if not opts.autoclose then
 		return
 	end
```
```suggestion
	if opts.autoclose then
		autocommand.add_autoclose()
	end
```

axkirillov commented on lua/hbac/setup.lua:
```diff
@@ -22,65 +79,24 @@ M.setup = function(opts)
-		end
-	})
-
+	M.setup_autoclose()
```
```suggestion

```

axkirillov commented on lua/hbac/setup.lua:
```diff
@@ -1,18 +1,75 @@
+
+			local buffers_to_close = num_buffers - M.opts.threshold
+
+			-- Buffer sorted by current > pinned > others (by bufnr)
```
~~Cool idea, would it be more performant if we sorted the list on every pin? 🤔 Perhaps nobody has that many open buffers for this to become a problem xD~~
Nevermind, what I wrote doesn't make any sense 😅

ditsuke replied:
I tested it with ~100 open buffers to no noticeable impact. I believe it shouldn't be a problem unless someone: disables autoclose, opens thousands of buffers and then enables autoclose or sets the buffer threshold absurdly high.

axkirillov commented on lua/hbac/setup.lua:
```diff
@@ -1,18 +1,75 @@
+				local windows_with_buf = vim.fn.win_findbuf(buf)
+				if name ~= "" and listed and #windows_with_buf == 0 then
+					return true
+				end
```
This is really cool! So if a a buffer is shown in any window, it will not get closed, if I understand correctly. This was a "bug" that I wanted to fix myself, but you beat me to it!

ditsuke replied:
Exactly, along with this and `close_command` (new configuration opt) we have all the tools to avoid unwanted behavior

axkirillov commented:
Since I rebased your other pr on top of main those 2 commits got rehashed. Just delete them and force push to get rid of conflicts.

Could you also update the README.md to mention the new toggle_autoclose command?

ditsuke committed:
refactor: Factor out autocommand setup to new mod

Also some features:
- Improved close-prioritising heuristic (incl some bug fixes)
- Configure to use custom close command (for example, `mini.bufremove`)

chore: Suggestion from code review

chore: Suggestions from code review

ditsuke committed:
chore(dev): Add selene and neoconf configs

- Selene is a linter for lua
- Neoconf is a project-wide configuration tool for lspconfig. It'll
  enable lua_ls with awareness for neovim APIs for plugin devs.

axkirillov commented on lua/hbac/command.lua:
```diff
@@ -1,26 +1,38 @@
 
+M.subcommands.toggle_autoclose = function()
+	if state.autoclose_enabled then
+		state.autoclose_enabled = false
+		autocommands.autoclose.disable()
+		print("Hbac: Autoclose disabled")
```
in this case it makes sense to check if pcall returns true and set the state after
```suggestion
		if autocommands.autoclose.disable() then
		    state.autoclose_enabled = false
		    print("Hbac: Autoclose disabled")
		end
```

ditsuke replied:
Couldn't the pcall failure indicate that the augroup didn't exist/was already removed by an external actor?

axkirillov replied:
Good point. That would be an implicit assumption though. Lets add a comment then.

ditsuke replied:
Absolutely, good idea.

axkirillov commented on lua/hbac/autocommands.lua:
```diff
@@ -0,0 +1,79 @@
+end
+
+M.autoclose.disable = function()
+	pcall(function()
```
```suggestion
	return pcall(function()
```

axkirillov commented on lua/hbac/setup.lua:
```diff
@@ -1,18 +1,24 @@
-
+	local command = require("hbac.command")
+	M.opts = vim.tbl_extend("force", M.opts, opts or {})
+	opts = M.opts
```
At this point we populated the M.opts will all correct values, so we should use that
```suggestion
```

axkirillov commented on lua/hbac/setup.lua:
```diff
@@ -26,64 +32,18 @@ M.setup = function(opts)
 
-	if not opts.autoclose then
-		return
+	if opts.autoclose then
```
```suggestion
	if M.opts.autoclose then
```

axkirillov commented on lua/hbac/command.lua:
```diff
@@ -1,26 +1,38 @@
 
+M.subcommands.toggle_autoclose = function()
+	if state.autoclose_enabled then
+		state.autoclose_enabled = false
+		autocommands.autoclose.disable()
```
something like that?
```suggestion
		pcall(autocommands.autoclose.disable())
		-- if the above command fails, the group probably didn't exist, revert the state to false
        state.autoclose_enabled = false
```

ditsuke replied:
Added this to autoclose.disable() instead.

axkirillov commented:
Looks great! @al-ce do you wanna take a look? I think it touches some of your code.

al-ce commented:
> Looks great! @al-ce do you wanna take a look? I think it touches some of your code.

Great PR! I tested it real quick for the sake of an extra eye, works well and I can't find any issues on a quick test. Don't know that I have anything more constructive to add so I'll just comment from a user perspective. The pin commands work as I would expect as when autoclosing is toggled off and then are toggled back on. I very much like the feature to prevent visible ~~windows~~ buffers from being auto closed.

I don't want this to hold up the  deployment but maybe in a future commit you could explain in the README why a user might want to customize the close_command. I can think of some creative use cases but maybe not what was intended for allowing this new option. Thanks for the new features @ditsuke !

axkirillov commented:
Do you wanna rebase anyhting? Otherwise, I'll just merge it as is

al-ce commented:
> Do you wanna rebase anyhting? Otherwise, I'll just merge it as is

I think this is a good commit history to keep as is in main since it shows the new features, modules, and config files added gradually (vs my PR which would've had a few too many bug fix commits to want to navigate through).

ditsuke commented:
Squashed down to 4 commits. Thanks for approving @axkirillov @al-ce :rocket:
